packages/web-forms/src/components/controls/RankControl.vue

```diff
@@ -0,0 +1,246 @@
+<script setup lang="ts">
+import ControlText from '@/components/ControlText.vue';
+import ValidationMessage from '@/components/ValidationMessage.vue';
+import type { RankNode } from '@getodk/xforms-engine';
+import type { Ref } from 'vue';
+import { computed, inject, ref } from 'vue';
+import { VueDraggable } from 'vue-draggable-plus';
+
+interface RankControlProps {
+	readonly question: RankNode;
+}
+
+interface HighlightOption {
+	index: Ref<number | null>;
+	timeoutID: NodeJS.Timeout | null;
+}
+
+const props = defineProps<RankControlProps>();
+const HOLD_DELAY = 200; // Delay in ms to hold an item before dragging, avoids accidental reordering on swipe.
+const touched = ref(false);
+const submitPressed = inject<boolean>('submitPressed');
+const disabled = computed(() => props.question.currentState.readonly === true);
+const highlight: HighlightOption = {
+	index: ref(null),
+	timeoutID: null,
+};
+
+const values = computed<string[]>({
+	get: () => {
+		const currentValues = props.question.currentState.value;
+
+		if (currentValues.length > 0) {
+			return currentValues.slice();
+		}
+
+		return props.question.currentState.valueOptions.map((option) => option.value);
+	},
+	set: (orderedValues) => {
+		touched.value = true;
+		props.question.setValues(orderedValues);
+	},
+});
+
+const setHighlight = (index: number | null) => {
+	highlight.index.value = index;
+
+	if (highlight.timeoutID) {
+		clearTimeout(highlight.timeoutID);
+		highlight.timeoutID = null;
+	}
+
+	if (highlight.index.value !== null) {
+		highlight.timeoutID = setTimeout(() => setHighlight(null), 1000);
+	}
+};
+
+const moveUp = (index: number) => {
+	const newPosition = index - 1;
+	if (newPosition < 0) {
+		return;
+	}
+	swapItems(index, newPosition);
+};
+
+const moveDown = (index: number) => {
+	const newPosition = index + 1;
+	if (newPosition >= values.value.length) {
+		return;
+	}
+	swapItems(index, newPosition);
+};
+
+const swapItems = (index: number, newPosition: number) => {
+	if (disabled.value) {
+		return;
+	}
+
+	setHighlight(index);
+
+	const swappedValues = values.value.slice();
+	const movedValue = swappedValues[index];
+
+	swappedValues[index] = swappedValues[newPosition];
+	swappedValues[newPosition] = movedValue;
+	values.value = swappedValues;
```

Comment:
This came out of some suggestions I made in an exploratory commit. Noting it because, if we find another use case, I'll want to extract a `@getodk/common` function for moving an array item to a different index. It would likely look more like the existing `insertAtIndex`, and probably call into it.

Reply:
Oh! I need to explore those libs. I'll keep it in mind next time, for now, leaving here.